### PR TITLE
Fix/mixup cutmix label dtype

### LINKED ISSUE #4657: RandomMixUpV2 / RandomCutMixV2 cast class labels to the image dtype: bfloat16 corrupts labels above 256
## Summary
`RandomMixUpV2` and `RandomCutMixV2` cast integer class labels to the **image dtype**. Under `bfloat16` (and `float16` above 2048) the labels themselves are rounded, so e.g. ImageNet-1k class ids come back wrong:

```python
import torch, kornia.augmentation as K
img = torch.rand(4, 1, 8, 8).bfloat16()
out, lab = K.RandomMixUpV2(p=1., data_keys=["input", "class"])(img, torch.tensor([257, 999, 2049, 4097]))
print(lab[:, 0].tolist())   # [256.0, 1000.0, 2048.0, 4096.0]
```
float16 gives `[257, 999, 2048, 4096]`. Same for CutMix.

## Expected
Labels should not share a lossy dtype with the image: keep the label columns in a dtype that represents them (e.g. float32/float64 floor, or return lambda separately).

Measured on `main` @ c643312a3 merged with #4634 (docs-only), torch 2.14.0, CPU, macOS arm64. Found while reviewing #4634 (batch 6d of #4407).

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by NiteeshGowda72**
## Description

Fixes #4657.

`RandomMixUpV2` and `RandomCutMixV2` previously converted class labels using the image/input dtype. This can corrupt class IDs when labels are represented with `float16` or `bfloat16`, especially for values above the exact-integer range of those dtypes.

This change ensures class labels use at least `float32` precision when the label dtype is not `float32` or `float64`, while preserving `float32` and `float64` inputs.

### Changes

- Prevent lossy `float16`/`bfloat16` conversion of MixUp class labels.
- Prevent lossy `float16`/`bfloat16` conversion of CutMix class labels.
- Preserve existing `float32` and `float64` behavior.
- Remove the now-unused `DType` import from `cutmix.py`.
- Add regression tests covering both `float16` and `bfloat16` with labels:
  `257, 999, 2049, 4097`.
- Add a changelog fragment.

### Validation

- Targeted regression tests: 4 passed
- Full `tests/augmentation/test_augmentation_mix.py`: 11622 passed, 3751 skipped, 44 xfailed, 11 xpassed
- `pixi run lint`: passed after Ruff auto-fix
- `pixi run typecheck`: no type errors; one unrelated existing deprecation warning
- `git diff --check`: clean

### CODE REVIEW

**Comment by NiteeshGowda72** [comment]
Hi @ducha-aiki, the latest CI run is showing 14 failing checks, including the CPU half-precision and float64 jobs. The targeted MotionBlur3D tests and local validation were passing on my side.

I’m not making further changes yet because I’d like to understand whether these CI failures are related to the current PR or to the broader test matrix/environment. Could you please take a look and let me know if you’d like me to investigate any specific failing job?

Thanks!

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Two P1 blockers remain at this head.

1. `mixup.py:120,136` and `cutmix.py:124,149` choose `calc_dtype` from the integer class-label tensor. Ordinary `int64` labels therefore force float32 even when the image/output contract is float64. This is the cause of the MixUp/CutMix `p=0` failures across the float64 matrix. Derive the computation/output dtype from the recorded image dtype and only promote float16/bfloat16 arithmetic where required.

2. Several changed CutMix expectations in `tests/augmentation/test_augmentation_mix.py` cast expected image tensors to float32 for half inputs. `out_image` correctly remains in the input image dtype, so the half jobs fail before the label contract is tested; the same-on-batch assertion still expects the original dtype as well. Keep image expectations in the parametrized image dtype and change only the label-output expectations.

The raw CI logs reproduce both issues consistently; these are code/test regressions rather than infrastructure failures.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5).

**Comment by NiteeshGowda72** [comment]
Hi @ducha-aiki, thanks for pointing out the two P1 issues. I’ve addressed both:
- calc_dtype is now derived from the recorded image dtype in params["dtype"] using DType.to_torch(...), so float64 image inputs continue to produce float64 label outputs while float16/bfloat16 are promoted to float32.
- Restored the CutMix image expectations to the parametrized image dtype and updated only the label-output expectations to use the promoted label dtype.
Validation after the changes:
- tests/augmentation/test_augmentation_mix.py: 62 passed
- Targeted MixUp/CutMix tests: 20 passed
- Ruff check/format: passed
- Typecheck: no type errors (only the existing custom_fwd deprecation warning)
- git diff --check: passed
The changes are pushed in commit da778fb8.
Could you please take another look when you have a chance?

**Review by ducha-aiki — APPROVED** [review]
Re-reviewed exact head a55e7140b1cb6211b0c89b105ae04e0f8b4ffd77.

Both prior dtype blockers are resolved. Calculation/output dtype is derived from the recorded image dtype, so float64 remains float64 while float16 and bfloat16 label arithmetic promotes to float32; CutMix image expectations are again asserted in the image dtype. Precision-sensitive values 257, 999, 2049, and 4097 survive the relevant paths. All 41 checks are green.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5).
